core/kernel/ldelf_syscalls.c

```diff
@@ -419,41 +432,26 @@ TEE_Result ldelf_set_prot(struct user_mode_ctx *uctx, uint32_t param_types,
 	return vm_set_prot(uctx, va, sz, prot);
 }
 
-TEE_Result ldelf_remap(struct user_mode_ctx *uctx, uint32_t param_types,
-		       TEE_Param params[TEE_NUM_PARAMS])
+TEE_Result ldelf_sc_remap(unsigned long old_va, vaddr_t *new_va,
+			  size_t num_bytes, size_t pad_begin, size_t pad_end)
 {
-	uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_INPUT,
-					  TEE_PARAM_TYPE_VALUE_INPUT,
-					  TEE_PARAM_TYPE_VALUE_INOUT,
-					  TEE_PARAM_TYPE_VALUE_INPUT);
 	TEE_Result res = TEE_SUCCESS;
-	uint32_t num_bytes = 0;
-	uint32_t pad_begin = 0;
+	struct ts_session *sess = ts_get_current_session();
+	struct user_mode_ctx *uctx = to_user_mode_ctx(sess->ctx);
 	uint32_t vm_flags = 0;
-	uint32_t pad_end = 0;
-	vaddr_t old_va = 0;
-	vaddr_t new_va = 0;
-
-	if (exp_pt != param_types)
-		return TEE_ERROR_BAD_PARAMETERS;
-
-	num_bytes = params[0].value.a;
-	old_va = reg_pair_to_64(params[1].value.a, params[1].value.b);
-	new_va = reg_pair_to_64(params[2].value.a, params[2].value.b);
-	pad_begin = params[3].value.a;
-	pad_end = params[3].value.b;
 
 	res = vm_get_flags(uctx, old_va, num_bytes, &vm_flags);
 	if (res)
 		return res;
 	if (vm_flags & VM_FLAG_PERMANENT)
 		return TEE_ERROR_ACCESS_DENIED;
 
-	res = vm_remap(uctx, &new_va, old_va, num_bytes, pad_begin,
-		       pad_end);
-	if (!res)
-		reg_pair_from_64(new_va, &params[2].value.a,
-				 &params[2].value.b);
+	res = vm_remap(uctx, new_va, old_va, num_bytes, pad_begin, pad_end);
 
 	return res;
 }
+
+TEE_Result ldelf_sc_gen_rnd_num(void *buf, size_t num_bytes)
+{
+	return crypto_rng_read(buf, num_bytes);
```

Comment:
Need to check that `buf` and size is safe to write at. Something like:
``` C
uint32_t flags = TEE_MEMORY_ACCESS_READ |
                 TEE_MEMORY_ACCESS_WRITE |
                 TEE_MEMORY_ACCESS_ANY_OWNER;
struct ts_session *sess = ts_get_current_session();
struct user_mode_ctx *uctx = to_user_mode_ctx(sess->ctx);
TEE_Result res = vm_check_access_rights(uctx, flags, buf, num_bytes);

if (res)
         return res;

Reply:
There are the `copy_from_user()` and friends helpers from **user_access.h** that could help.
They need to be upgraded from `to_user_ta_ctx()->utcx` to `to_user_mode_ctx()`

Reply:
@jenswi-linaro: Thanks, good catch. `utee_param_to_param()` was taking care of the access check for all memref type parameters. Since that's not used here, I think we also need the checks in `ldelf_sc_open_bin()` and `ldelf_sc_copy_from_bin()` right?

@etienne-lms: Thanks, I'll take a look.

Reply:
> @jenswi-linaro: Thanks, good catch. `utee_param_to_param()` was taking care of the access check for all memref type parameters. Since that's not used here, I think we also need the checks in `ldelf_sc_open_bin()` and `ldelf_sc_copy_from_bin()` right?

Correct, I forgot to mention that.

Reply:
Done